File: android/app/src/main/java/com/example/curier_mobile/domain/model/Order.kt
package com.example.curier_mobile.domain.model

/**
 * Domain model для заказа курьера.
 *
 * Согласуется с `OrderCourierResponse` бэкенда (см. `docs/orders.md`),
 * поэтому здесь нет поля `price` — оно admin-only.
 */
data class Order(
    val id: String,
    val orderNumber: String,
    val customerName: String,
    val customerPhone: String?,
    val deliveryAddress: String,
    val productDescription: String?,
    val comments: String?,
    val status: OrderStatus,
    val priority: OrderPriority = OrderPriority.NORMAL,
    val courierId: String?,
    val createdAt: String,
    val assignedAt: String?,
    val pickedUpAt: String?,
    val nearCustomerAt: String?,
    val deliveredAt: String?,
    val returnedAt: String?,
    val photos: List<Photo> = emptyList()
)

/**
 * Order status enum. Совпадает с `OrderStatus` на бэке.
 *
 * `NEW` присутствует для полноты, но курьер видит только заказы
 * в `assigned+` (бэк-фильтр в `GET /api/courier/orders/active`).
 */
enum class OrderStatus(val value: String, val displayName: String) {
    NEW("new", "Новый"),
    ASSIGNED("assigned", "Назначен"),
    PICKED_UP("picked_up", "Забрал заказ"),
    NEAR_CUSTOMER("near_customer", "Возле дома клиента"),
    DELIVERED("delivered", "Передал заказ"),
    RETURNED("returned", "Вернулся на предприятие");

    companion object {
        fun fromValue(value: String): OrderStatus {
            return entries.find { it.value == value }
                ?: throw IllegalArgumentException("Unknown status: $value")
        }

        /**
         * Single-step transition allowed for couriers.
         * `NEW` -> `ASSIGNED` происходит только на бэке (auto-assign);
         * курьер не может его двигать.
         */
        fun getNextStatus(current: OrderStatus): OrderStatus? {
            return when (current) {
                NEW -> null
                ASSIGNED -> PICKED_UP
                PICKED_UP -> NEAR_CUSTOMER
                NEAR_CUSTOMER -> DELIVERED
                DELIVERED -> RETURNED
                RETURNED -> null
            }
        }

        fun isValidTransition(from: OrderStatus, to: OrderStatus): Boolean {
            return getNextStatus(from) == to
        }
    }
}

/**
 * Приоритет заказа. Совпадает с `OrderPriority` на бэке. Курьер видит его
 * бэйджем — срочные заказы бэк назначает быстрым/опытным курьерам и
 * разбирает из очереди первыми (см. docs/assignment.md).
 */
enum class OrderPriority(val value: String, val displayName: String) {
    LOW("low", "Низкий"),
    NORMAL("normal", "Обычный"),
    HIGH("high", "Срочный");

    companion object {
        /** Неизвестное/отсутствующее значение трактуем как NORMAL — приоритет
         *  не критичен для работы курьера, падать из-за него не нужно. */
        fun fromValue(value: String?): OrderPriority =
            entries.find { it.value == value } ?: NORMAL
    }
}

File: admin/src/types/order.ts
/**
 * Доменные типы заказа.
 *
 * Соответствуют §4 (БД) и §5 (admin DTO) из Documentation/completion_plan.md.
 * На Этапе 1 используются для типизации моков; на Этапе 3 будут совпадать
 * с DTO бэкенда (UUID-идентификаторы, ISO-даты).
 */

export type OrderStatus =
    | "new"
    | "assigned"
    | "picked_up"
    | "near_customer"
    | "delivered"
    | "returned";

/**
 * Приоритет (срочность) заказа. Управляет взвешенным авто-назначением и
 * порядком разбора очереди на бэкенде (§8, docs/assignment.md).
 */
export type OrderPriority = "low" | "normal" | "high";

export const ORDER_PRIORITY_LABELS: Record<OrderPriority, string> = {
    low: "Низкий",
    normal: "Обычный",
    high: "Срочный",
};

/**
 * Подпись бэйджа статуса заказа в админке.
 *
 * Заметка про `delivered` vs `returned`: на бэкенде это два разных состояния
 * (см. `docs/orders.md` → «Status flow»). `delivered` — курьер передал заказ
 * клиенту; `returned` — курьер вернулся на базу, заказ закрыт. С точки зрения
 * админа оба — «заказ доставлен», поэтому подпись одинаковая. Точные моменты
 * перехода видны в timeline-секции drawer-а заказа.
 */
export const ORDER_STATUS_LABELS: Record<OrderStatus, string> = {
    new: "Новый",
    assigned: "Назначен",
    picked_up: "Забран",
    near_customer: "Рядом с клиентом",
    delivered: "Доставлен",
    returned: "Доставлен",
};

/** Активные статусы — заказ в работе (виден в дефолтном фильтре). */
export const ACTIVE_ORDER_STATUSES: ReadonlySet<OrderStatus> = new Set([
    "new",
    "assigned",
    "picked_up",
    "near_customer",
]);

/**
 * Метаданные фотографии доставки. Bytes тянутся через
 * `GET /api/admin/orders/:id/photo/:photoId` (см. docs/photos.md).
 *
 * `expiresAt` — момент удаления файла cron-ом TTL (см. settings.md).
 */
export interface PhotoMeta {
    id: string;
    uploadedAt: string;
    expiresAt: string;
}

export interface Order {
    id: string;
    orderNumber: string;
    customerName: string;
    customerPhone: string;
    deliveryAddress: string;
    productDescription: string;
    comments: string | null;
    /** Курьер не видит цену, админ видит. На Этапе 1 — мок. */
    price: number | null;
    status: OrderStatus;
    priority: OrderPriority;
    courierId: string | null;
    createdByAdminId: string;
    createdAt: string;
    assignedAt: string | null;
    pickedUpAt: string | null;
    nearCustomerAt: string | null;
    deliveredAt: string | null;
    returnedAt: string | null;
    /**
     * Метаданные фото. На list-эндпоинтах backend всегда отдаёт `[]`
     * (см. docs/photos.md): реальный список приходит только на
     * `GET /api/admin/orders/:id` и transition/reassign-ответах.
     */
    photos: PhotoMeta[];
}

/**
 * ISO-метка момента, когда заказ вошёл в свой текущий статус. Берётся из
 * соответствующего timestamp-поля; для `new` — `createdAt`. Используется,
 * чтобы показать «сколько времени заказ в текущем статусе» — счётчик
 * сбрасывается при каждом переходе, потому что timestamp у нового статуса
 * стампится бэкендом отдельным полем.
 */
export function getOrderStatusSince(order: Order): string {
    switch (order.status) {
        case "new":
            return order.createdAt;
        case "assigned":
            return order.assignedAt ?? order.createdAt;
        case "picked_up":
            return order.pickedUpAt ?? order.createdAt;
        case "near_customer":
            return order.nearCustomerAt ?? order.createdAt;
        case "delivered":
            return order.deliveredAt ?? order.createdAt;
        case "returned":
            return order.returnedAt ?? order.createdAt;
    }
}
